### PR TITLE
Exercise fuzzers during CI builds

### PR DESCRIPTION
**Pull request by ilammy**
We have had some kind of a fuzzing framework for close to a year, but it did not get enough attention since only nobody cares about that. Let’s improve the situation a bit.

- Make sure that fuzzers can be built with sanitizers. This is the most interesting case since sanitizers report bugs by crashing and AFL reacts to crashes. Building and running ASAN tests was a quest on its own.

- Build and run fuzzers on CI. Every time. For 30 seconds each. Hopefully we’ll catch some low-hanging bugs with that, given a sufficient number of builds.

- There is at least one known bug which is caught by fuzzers+sanitizers, so hopefully the build will fail this time and you can see the report.

- Improve reporting and crash analysis tool so that they are more useful.

There are more implementation details in commit messages.

<details>
<summary>P.S. Yep, the build <a href="https://app.circleci.com/jobs/github/cossacklabs/themis/7683">does fail</a> right now.</summary>

```
fuzzing scell_seal_roundtrip...
afl-fuzz 2.52b by <lcamtuf@google.com>
[+] Looks like we're not running on a tty, so I'll be a bit less verbose.
[+] You have 36 CPU cores and 19 runnable tasks (utilization: 53%).
[+] Try parallel jobs - see /usr/share/doc/afl-doc/docs/parallel_fuzzing.txt.
[*] Checking CPU core loadout...
[+] Found a free CPU core, binding to #0.
[*] Checking core_pattern...
[*] Setting up output directories...
[+] Output directory exists but deemed OK to reuse.
[*] Deleting old session data...
[+] Output dir cleanup successful.
[*] Scanning 'tools/afl/input/scell_seal_roundtrip'...
[+] No auto-generated dictionary tokens to reuse.
[*] Creating hard links for all input files...
[*] Validating target binary...
[*] Attempting dry run with 'id:000000,orig:ok_short_message.dat'...
[*] Spinning up the fork server...
[+] All right - fork server is up.
    len = 30, map size = 314, exec speed = 3706 us
[+] All test cases processed.

[+] Here are some useful stats:

    Test case count : 1 favored, 0 variable, 1 total
       Bitmap range : 314 to 314 bits (average: 314.00 bits)
        Exec timing : 3706 to 3706 us (average: 3706 us)

[*] No -t option specified, so I'll use exec timeout of 20 ms.
[+] All set and ready to roll!
[*] Entering queue cycle 1.
[*] Fuzzing test case #0 (1 total, 0 uniq crashes found)...
[*] Fuzzing test case #1 (8 total, 0 uniq crashes found)...
[*] Fuzzing test case #3 (8 total, 0 uniq crashes found)...
[*] Fuzzing test case #4 (8 total, 0 uniq crashes found)...
[*] Fuzzing test case #6 (8 total, 0 uniq crashes found)...


+++ Testing aborted by user +++
[+] We're done here. Have a nice day!

see build/afl/output/scell_seal_roundtrip/2020-01-27_08-00-03 for results

Analyzing results...

# scell_seal_decrypt -- 2020-01-27_07-59-33

## id:000000 -- SIGABRT

Run:

    build/afl/scell_seal_decrypt build/afl/output/scell_seal_decrypt/2020-01-27_07-59-33/crashes/id:000000,sig:06,src:000000,op:flip32,pos:27

Input (base64):

AAAAA2tleQAAAAdjb250ZXh0AAAAMwABAUAM////7wAAAAcAAACbMpUEngYcUNl6
KWhBWxWP8Qp1AC4xdvnfM1NJYB5sxBGHsA==

Debugger output:

=================================================================
==3989==ERROR: AddressSanitizer: heap-buffer-overflow on address 0xf60015bc at pc 0xf7a1e8be bp 0xff8a27f8 sp 0xff8a23c8
READ of size 16 at 0xf60015bc thread T0
    #0 0xf7a1e8bd  (/usr/lib32/libasan.so.4+0x778bd)
    #1 0xf7813a23  (/usr/lib/i386-linux-gnu/libcrypto.so.1.1+0x118a23)
    #2 0xf781f05e in EVP_CIPHER_CTX_ctrl (/usr/lib/i386-linux-gnu/libcrypto.so.1.1+0x12405e)
    #3 0x56659e62 in soter_sym_aead_decrypt_final src/soter/openssl/soter_sym.c:433
    #4 0x5664dcb9 in themis_auth_sym_plain_decrypt src/themis/sym_enc_message.c:109
    #5 0x5664dcb9 in themis_auth_sym_decrypt_message_ src/themis/sym_enc_message.c:332
    #6 0x5664dcb9 in themis_auth_sym_decrypt_message src/themis/sym_enc_message.c:439
    #7 0x566490ad in themis_secure_cell_decrypt_seal src/themis/secure_cell.c:92
    #8 0x56646c12 in main tools/afl/src/scell_seal_decrypt.c:99
    #9 0xf7537e80 in __libc_start_main (/lib/i386-linux-gnu/libc.so.6+0x18e80)
    #10 0x56647a82  (/home/user/project/build/afl/scell_seal_decrypt+0x3a82)

Address 0xf60015bc is a wild pointer.
SUMMARY: AddressSanitizer: heap-buffer-overflow (/usr/lib32/libasan.so.4+0x778bd) 
Shadow bytes around the buggy address:
  0x3ec00260: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x3ec00270: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x3ec00280: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x3ec00290: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x3ec002a0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
=>0x3ec002b0: fa fa fa fa fa fa fa[fa]fa fa fa fa fa fa fa fa
  0x3ec002c0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x3ec002d0: fa fa fa fa 00 00 00 00 00 00 03 fa fa fa fa fa
  0x3ec002e0: 00 00 00 00 00 00 00 01 fa fa fa fa fa fa fa fa
  0x3ec002f0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x3ec00300: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==3989==ABORTING

Exited with code exit status 1
```

</details>

Failures are also stored in artifacts, [example](https://app.circleci.com/jobs/github/cossacklabs/themis/7732/artifacts).

## Checklist

- [X] Change is covered by automated tests
- [X] The [coding guidelines] are followed
- [X] Public API has proper documentation
- [X] Example projects and code samples are updated
- [X] ~~Changelog is updated~~ (maybe later)
- [x] Parsing bug is fixed and the build is not red

[coding guidelines]: https://github.com/cossacklabs/themis/blob/master/CONTRIBUTING.md

### CODE REVIEW

File: tools/afl/src/readline.c [hunk 1]
```diff
@@ -16,21 +16,28 @@
 
 #include "readline.h"
 
+#include <errno.h>
 #include <stdlib.h>
 
+#define MAX_SANE_LENGTH (50 * 1024 * 1024)
```
Comment (vixentael) [inline tools/afl/src/readline.c]: max 50 MBytes, right?

**Comment by vixentael** [comment]
So,

1. Every CI check is 30sec slower.
2. If fuzzer finds smth, CI tests will fail.
3. What is our procedure on fixing things that fuzzer found? I mean, what PR author should do, if fuzzer found an issue that is not related to their PR (most things won't be related).

**Comment by ilammy** [comment]
@Lagovas,

> soon Bash and Makefile languages will be at top 5 of Themis languages on github )

YAML programmers 🤣 

@vixentael,

> Every CI check is 30sec slower.

They are running in parallel, and most of the time Android builds and integration tests are the slowest, so we're unlikely to be blocked by this. (And it will be about 2 minutes and counting in fact: there are 2 fuzzing tools and 2 fuzzing modes, each of those 4 runs for 30 seconds.)

> If fuzzer finds smth, CI tests will fail.

That's correct.

> What is our procedure on fixing things that fuzzer found?

I guess manually inspect the failure and use our best judgement. I'd probably rerun the job, if the failure persists and is definitely not related to PR, they we might note the failure, approve the merge with it, and fix it later separately.

Best practice would be to run shadow builds and send alerts about fuzzing failures only to maintainers, without exposing PR authors to them. But I'm not really sure how this can be easily done with CircleCI. I did not notice any ready-made so we'd have to hack something (e.g., note the failure, send an alert via API integration, let the build succeed if the failure is found by fuzzers).

File: tools/afl/src/readline.c (hunk 1 above)
Comment (ilammy) [reply tools/afl/src/readline.c]: Yeah, 50 MB. This is default memory limit of AFL and our test data is unlikely to exceed it. There is no gain in making a huge test case as AFL will try minimizing it anyway.

**Comment by Lagovas** [comment]
> I guess manually inspect the failure and use our best judgement. I'd probably rerun the job, if the failure persists and is definitely not related to PR, they we might note the failure, approve the merge with it, and fix it later separately.

does fuzzer can save test data to some file? if so we can save it as test artifacts to reproduce it later on failures

**Comment by ilammy** [comment]
> does fuzzer can save test data to some file?

Yes. The results are saved in the build tree, along with all the intermediate data. We use this data to generate reports (which are dumped into build logs). There should be enough data in the logs to reproduce the crashes:

- what mutation AFL applied (e.g., flipped 1 bit at offset 27)
- name of the tool that failed
- full input to the tool
- backtrace & register data, if available
- anything else printed by the sanitizer, usually pretty detailed description

We do not save the original report data since I thought all of the above should be sufficient. But it should be easy to store full reports on CircleCI as builds artifacts. Who's for it?

**Comment by ilammy** [comment]
[Artifacts seem to work now](https://app.circleci.com/jobs/github/cossacklabs/themis/7732/artifacts).

File: tools/afl/src/readline.c (hunk 1 above)
Comment (vixentael) [reply tools/afl/src/readline.c]: makes sense, thanx!

**Comment by vixentael** [comment]
In my perfect world (🦄) we'd have some artifact (logs in file, result of AFL) that we can grab from CI and post into internal issue tracker. So, smth that can be downloaded and attached.

Artifacts should work i think

**Comment by ilammy** [comment]
PR #592 contains fixes which should resolve issues found by AFL in this PR. Once that is merged, **master** should be merged into this PR and the build is expected to go green.

**Pull request merged**